src/Featurizers/Components/VectorNormsEstimator.h

```diff
@@ -0,0 +1,167 @@
+    static constexpr char const * const     NameValue = VectorNormsEstimatorName;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+    VectorNormsTrainingOnlyPolicy(void);
+
+    void fit(InputType const &input);
+    VectorNormsAnnotationData<UpdaterType> complete_training(void);
+
+private:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Types
+    // |
+    // ----------------------------------------------------------------------
+    std::vector<std::double_t> _norms;
+    UpdaterType _updater;
+};
+
+} // namespace Details
+
+/////////////////////////////////////////////////////////////////////////
+///  \typedef       VectorNormsEstimator
+///  \brief         A training-only class that finds vector-based norms of a matrix
+///
+template <
+    typename InputT,
+    typename UpdaterType,
+    size_t MaxNumTrainingItemsV=std::numeric_limits<size_t>::max()
+>
+using VectorNormsEstimator                       = TrainingOnlyEstimatorImpl<Details::VectorNormsTrainingOnlyPolicy<InputT, UpdaterType>, MaxNumTrainingItemsV>;
+
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// |
+// |  Implementation
+// |
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+
+// ----------------------------------------------------------------------
+// |
+// |  VectorNormsAnnotationData
+// |
+// ----------------------------------------------------------------------
+template <typename UpdaterType>
+VectorNormsAnnotationData<UpdaterType>::VectorNormsAnnotationData(std::vector<std::double_t> norms) :
+    Norms(std::move(norms)) {
+        for (auto const & norm : Norms) {
+            if (norm < 0) {
+                throw std::invalid_argument("Norms shouldn't be less than 0!");
+            }
+        }
+}
+
+// ----------------------------------------------------------------------
+// |
+// |  Details::VectorNormsTrainingOnlyPolicy
+// |
+// ----------------------------------------------------------------------
+template <typename T, typename UpdaterType>
+Details::VectorNormsTrainingOnlyPolicy<T, UpdaterType>::VectorNormsTrainingOnlyPolicy(void) {
+}
+
+template <typename T, typename UpdaterType>
+void Details::VectorNormsTrainingOnlyPolicy<T, UpdaterType>::fit(InputType const &input) {
+    // T is a pair of iterators
+    // unpack input to two iterators
+    IteratorType & begin = const_cast<IteratorType&>(std::get<0>(input));

```

Comment:
>const_cast<IteratorType&> [](start = 27, length = 25)

This const_cast shouldn't be necessary. If you need to use the value for iteration, make a copy rather than taking a reference.